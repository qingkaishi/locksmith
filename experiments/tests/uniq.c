/* Basic uniqueness test on a local variable; accesses to the variable
   before it becomes shared are legal. */
#include <pthread.h>
#include <stdio.h>

int *j;
pthread_mutex_t l;

void * malloc(int);

void * dostuff(void * a) {
  int *k;
  int **m;
  
  k = (int*) malloc(sizeof (int));
  *k = 2;
  pthread_mutex_lock(&l);
  j = k;
  *j = 3;
  pthread_mutex_unlock(&l);
  return NULL;
}

int main() {
  pthread_t t1;
  pthread_mutex_init(&l, NULL);
  int i = 1;
  pthread_create(&t1, NULL, dostuff, NULL);
  pthread_create(&t1, NULL, dostuff, NULL);
}


