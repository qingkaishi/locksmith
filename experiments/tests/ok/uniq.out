************************* STARTING *************************

shared: &j:uniq.c:6
        alloc:uniq.c:15
&j:uniq.c:6 is protected by:
  concrete l:uniq.c:26
  l:uniq.c:7
alloc:uniq.c:15 is protected by:
  concrete l:uniq.c:26
  l:uniq.c:7
*************************** DONE ***************************
